Sources/PackageCollections/Storage/SQLitePackageCollectionsStorage.swift

```diff
@@ -557,92 +561,84 @@ final class SQLitePackageCollectionsStorage: PackageCollectionsStorage, Closable
-                                        map[collection.identifier] = entry
+                                version.manifests.values.forEach { manifest in
+                                    manifest.targets.forEach { target in
+                                        let match: Bool
+                                        switch type {
+                                        case .exactMatch:
+                                            match = target.name.lowercased() == query
+                                        case .prefix:
+                                            match = target.name.lowercased().hasPrefix(query)
+                                        }
+                                        if match {
+                                            // Avoid copy-on-write: remove entry from dictionary before mutating
+                                            var entry = map.removeValue(forKey: collection.identifier) ?? .init()
+                                            entry.append((target, package))
+                                            map[collection.identifier] = entry
+                                        }
                                     }
                                 }
                             }
                         }
                         return map
                     }
 
-                    // compose result :p
-                    var packageCollections = [PackageReference: (package: Model.Package, collections: Set<Model.CollectionIdentifier>)]()
-                    var targetsPackages = [Model.Target: Set<PackageReference>]()
-
                     collectionsPackages.forEach { collectionIdentifier, packagesAndTargets in
                         packagesAndTargets.forEach { item in
                             // Avoid copy-on-write: remove entry from dictionary before mutating
-                            var packageCollectionsEntry = packageCollections.removeValue(forKey: item.package.reference) ?? (item.package, .init())
+                            var packageCollectionsEntry = packageCollections.removeValue(forKey: item.package.reference.identity) ?? (item.package, .init())
                             packageCollectionsEntry.collections.insert(collectionIdentifier)
-                            packageCollections[item.package.reference] = packageCollectionsEntry
-
-                            // Avoid copy-on-write: remove entry from dictionary before mutating
-                            var targetsPackagesEntry = targetsPackages.removeValue(forKey: item.target) ?? .init()
-                            targetsPackagesEntry.insert(item.package.reference)
-                            targetsPackages[item.target] = targetsPackagesEntry
+                            packageCollections[item.package.reference.identity] = packageCollectionsEntry
+
+                            packageCollectionsEntry.package.versions.forEach { version in
```

Comment:
I rewrote the logic in this part such that the logic for processing result becomes the same for both code all paths.